### ISSUE TITLE
Investigate Perf Drop for fairseq gpu training on aws batch with EFA (p3dn)

### ISSUE THREAD

**Issue by kiukchung** [pytorch/torchx#444]
## 🐛 Bug

Not really a TorchX bug per-se, but opening this issue to keep track of this investigation. 

There seems to be a perf drop when running the same fairseq trainer on AWS Batch on p3dn (w/ EFA) versus running on the same host baremetal (no Docker).

Module (check all that applies):
 * [ ] `torchx.spec`
 * [ ] `torchx.component`
 * [ ] `torchx.apps`
 * [ ] `torchx.runtime`
 * [ ] `torchx.cli`
 * [ ] `torchx.schedulers`
 * [ ] `torchx.pipelines`
 * [ ] `torchx.aws`
 * [ ] `torchx.examples`
 * [x] `other`

Below is a table summarizing the different runs:

Common Params:
1. Instance Type: p3dn.24xlarge
2. Num Nodes: 2
3. Nodes in same placement group: Yes
4. Environment Variables:
     a. LOGLEVEL=INFO
     b. NCCL_SOCKET_IFNAME=eth0
     c. NCCL_ALGO=RING
     d. NCCL_PROTO=simple 
     e. FI_PROVIDER=efa
     f. FI_EFA_USE_DEVICE_RDMA=1
 5. In all cases RDMA is not enabled (seems to not be available on p3dn see: https://github.com/aws/aws-ofi-nccl/issues/104)
 
|Scheduler     | Docker       | EFA picked up by NCCL |  NCCL_ALGO | Performance | Log                |
|-------------|:----------- |:-------------------------|:-------------|:-------------|:-------------|
| BareMetal    | No               | No                                      | Tree      | *baseline* (~110k wpm)  | [gist-link](https://gist.github.com/kiukchung/230e1bc13d17ac275f5b053ff12d6534)| 
| BareMetal    | Yes              | Yes                                     | Tree     | WIP               |                        |  
| AWS Batch  | Yes              | Yes                                     | Ring              | ~44k wpm      |                        |
| AWS Batch  | Yes              | Yes                                     | Tree              | ~70k wpm

* Scheduler BareMetal: ssh onto the nodes and kick off the trainer manually
* Docker Yes/No: trainer runs in docker (e.g. `docker run`) versus `python -m torch.distributed.launch`)

## To Reproduce

1.
1.
1.

## Environment

 - torchx version (e.g. 0.1.0rc1): torchx-nightly
 - Python version: 3.8+
 - OS (e.g., Linux): Amazon Linux2
 - How you installed torchx (`conda`, `pip`, source, `docker`): pip
 - Docker image and tag (if using docker): 
 - Git commit (if installed from source): N/A
 - Execution environment (on-prem, AWS, GCP, Azure etc): AWS Batch
 - Any other relevant information:

**Comment by nzmsv** [comment]
Quick sanity check: are you allowing the container access to the EFA device?  Something like `docker run --device=/dev/infiniband/uverbs0` or `docker run --privileged`?

**Comment by kiukchung** [comment]
@yukunlin can provide more details (as he is running the benchmarks from Amazon Search side). The benchmarks were run via AWS Batch and yes, we are setting the equivalent of `--device=/dev/infiniband/uverbs0` by setting the `linuxParameters.devices` field in the Batch JobDefinition (https://docs.aws.amazon.com/batch/latest/APIReference/API_LinuxParameters.html). See https://github.com/pytorch/torchx/blob/main/torchx/schedulers/aws_batch_scheduler.py#L170.

**Comment by nzmsv** [comment]
Could you get a run with
```
    -e NCCL_DEBUG=INFO \
    -e FI_LOG_LEVEL=Info \
```
both with and without Docker to help determine what the difference is?

**Comment by rashikakheria** [comment]
Also run mpirun with `--allow-run-as-root`. Please see https://github.com/aws-samples/aws-efa-eks/blob/main/examples/pytorch/fairseq1B-GPT3-efa-example.yaml for more details.

**Comment by kiukchung** [comment]
@rashikakheria  we are running with [`torch.distributed.run`](https://pytorch.org/docs/stable/elastic/run.html) which is the builtin pytorch-distributed launcher (so not using mpi). Are there any perf implications (e.g. not being able to load drivers and such) when not running as root?

**Comment by yukunlin** [comment]
Posting some details on the setup I've been using for testing.

## Repro steps
Steps to reproduce observed behavior:

1. Set up batch training environment (see "Environment" section)
1. Follow https://github.com/pytorch/fairseq/tree/main/examples/language_model to download and pre-process the data
1. In the torchx workspace, create a wrapper for `fairseq_train_wrapped`
   
   ```python
	#!/opt/conda/bin/python3.8
	# -*- coding: utf-8 -*-
	import os
	import sys
	from fairseq_cli.train import cli_main
	if __name__ == "__main__":
	    sys.argv += ["--local_rank", os.getenv("LOCAL_RANK")]
	    sys.exit(cli_main())
   ```
1. Kick off a 2 node training job:

	```sh
	torchx run  --scheduler aws_batch --scheduler_args queue=fairseq-multi-node,image_repo="${ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com/yukunlin-test" dist.ddp --image "${ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com/yukunlin-test-snap" --env NCCL_SOCKET_IFNAME=eth0,LOGLEVEL=INFO,FI_PROVIDER=efa,NCCL_ALGO=TREE,NCCL_PROTO=simple --gpu 8 --memMB 720000 --j 2x8 --mounts type=bind,src=/mnt/fsx,dst=/job,type=device,src=/dev/infiniband/uverbs0 --script fairseq_train_wrapped -- --task language_modeling \
	  /job/fairseq/data-bin/wikitext-103 \
	  --save-dir "/job/fairseq/checkpoints/transformer_wikitext-103_m24_$(date +%s)" \
	  --arch transformer_lm --share-decoder-input-output-embed \
	  --dropout 0.1 \
	  --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0 \
	  --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
	  --tokens-per-sample 512 --sample-break-mode none \
	  --max-tokens 2048 \
	  --max-update 50000 \
	  --log-interval 30 --log-format json
	```
1. Change the `--j` parameter to `--j 1x8` to kick off a single node training job.
	
## Expected behavior

We'd expect that the 2 node training job to be faster than the single node job. For `fairseq-train`, throughput is measured by `wpm` (words per minute). For single node we observe ~70K wpm. However, on 2 nodes, we observe ~66K wpm.

On a setup where the same `fairseq-train` job is run on bare EC2 (no docker), single node throughput is also ~70K wpm, but 2 node's throughput is ~110K wpm.


## Environment

* torchx version: `torchx-nightly-2022.4.4`
* Python version: 3.8+
* OS (e.g., Linux): Ubuntu 20.04 for docker image, Amazon Linux 2 for EC2 host
* Docker image and tag: Using [the AWS DL pytorch 1.10 training container as base image](https://github.com/aws/deep-learning-containers/blob/master/pytorch/training/docker/1.10/py3/cu113/Dockerfile.e3.gpu)
  Dockerfile:
  
  ```
  FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:1.10.2-gpu-py38-cu113-ubuntu20.04-e3

  RUN pip install subword_nmt tqdm boto3 awscli jellyfish fairseq tensorboardX
  ```
* Execution environment: 
	* AWS Batch:
		* AMI: `amazon/amzn2-ami-ecs-gpu-hvm-2.0.20220318-x86_64-ebs` (`ami-076a815495dfc0dd1`)
		* Region: `us-west-2`
		* Data store: [FSx for Lustre](https://aws.amazon.com/fsx/lustre/), bind mount to docker
		* Instance type: `p3dn.24xlarge`, EFA enabled in same placement group (cluster)

**Comment by nzmsv** [comment]
Keep in mind that PyTorch is... weird.  By default it will build its own copy of almost everything, including NCCL and CUDA.  It's possible that you are using a different version of NCCL than you bargained for.

**Comment by rashikakheria** [comment]
The best way to debug this performance drop is to enable `NCCL_DEBUG=info` which would let us know if EFA was being used or not (as Greg suggested earlier).

**Comment by yukunlin** [comment]
Yes, EFA was being used according to the logs:

```
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO NCCL_SOCKET_IFNAME set by environment to eth0
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO NCCL_SOCKET_IFNAME set to eth0
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO Bootstrap : Using eth0:10.0.0.159<0>
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO NET/Plugin: Failed to find ncclCollNetPlugin_v4 symbol.
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO NET/OFI Forcing AWS OFI ndev 4
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO NET/OFI Selected Provider is efa
[0]:ip-10-0-0-159:24:24 [0] NCCL INFO Using network AWS Libfabric
[4]:ip-10-0-0-159:28:28 [4] NCCL INFO NCCL_SOCKET_IFNAME set by environment to eth0
[2]:ip-10-0-0-159:26:26 [2] NCCL INFO NCCL_SOCKET_IFNAME set by environment to eth0
[5]:ip-10-0-0-159:29:29 [5] NCCL INFO NCCL_SOCKET_IFNAME set by environment to eth0
[5]:ip-10-0-0-159:29:29 [5] NCCL INFO NCCL_SOCKET_IFNAME set to eth0
[2]:ip-10-0-0-159:26:26 [2] NCCL INFO NCCL_SOCKET_IFNAME set to eth0
```

**Comment by rashikakheria** [comment]
@yukunlin Could you provide access to complete logs?

**Comment by yukunlin** [comment]
@rashikakheri complete logs here: https://gist.github.com/yukunlin/fb068d873950a6a3519db9d475fc54e6

**Comment by nzmsv** [comment]
@yukunlin : I'm not seeing a version printout message from the OFI plugin in the logs above, which suggests that it's a pretty old version.  Which version is being used?  Is it possible to upgrade?

**Comment by yukunlin** [comment]
Here are logs with a version print out of the OFI version (`Using aws-ofi-nccl 1.2.0aws`): https://gist.github.com/yukunlin/634c600a11e36d1384215ab08366e774

I detailed more my investigation here: https://github.com/aws/aws-ofi-nccl/issues/106. The TLDR is that the discrepancy is between running on AL2 vs Ubuntu (AWS vended deep learning AMIs). I think we can close this here and move to https://github.com/aws/aws-ofi-nccl/issues/106 instead.
